Improve the performance of the type loader through various tweaks

davidwrighton opened the pull request:
1. Disable type validation by default in ReadyToRun images
   - This is potentially a behavior changing modification, and so maybe we shouldn't do this; however it produces some nice wins
   - I've built a type validation layer into the compiler to validate types against (most of) the checks that are removed. As part of building this, I've found a few small bugs in the typesystem. This validation layer isn't quite complete, and will be optional, although it will be enabled by default in crossgen2.
2. Add new metadata to R2R images for finding generic parameter and nested class information
   - These lookups are extremely common during type loading.
3. Store the index of the MethodDesc within the MethodDescChunk as well as the chunk index.
   - Used to improve the performance of finding the temporary entry point, which is a surprisingly common operation during type load. Its converted from an O(N^2) algorithm to O(1)
4. Remove an old ngen'ism (CanShareVTableChunk) Since we don't have ngen images, we can ALWAYS share vtable chunks

Overall, this set of changes produces the following performance impact

Powershell startup

Initial time to launch powershell 100 times[^1] - 22869ms
With all fixes enabled - 19966ms
Without enabling skipping type validation - 21240ms

Additionally, while this change does increase the size of the r2r compiled files, the total size of the dlls associated with powershell goes from 242,357,432 bytes to 242,564,792 bytes, which is a fairly minimal increase.

[^1]:  These powershell startup numbers are done by taking an official build of powershell targetting an older version of .NET, replacing the version of .NET with a locally built one, and then running crossgen2 on all binaries of powershell. This isn't likely how Powershell actually produces its builds, but should be representative of the sort of performance wins that can be had with these changes.

davidwrighton committed:
Skip type validation by default in ReadyToRun images
- Technically, this is a breaking change, so I've provided a means for disabling the type validation skip
- The model is that the C# compile won't get these details wrong, so disable the checks when run through crossgen2. The idea is that we'll get these things checked during normal, non-R2R usage of the app, and publish won't check these details.

davidwrighton committed:
Replace expensive lookups of generic parameter and nested class data with R2R optimized forms

davidwrighton committed:
Store index of MethodDesc as well as ChunkIndex. Makes MethodDesc::GetTemporaryEntryPoint *much* faster

davidwrighton committed:
Optimize the path for computing if a method is eligible for tiered compilation

davidwrighton committed:
Remove CanShareVtableChunksFrom concept
- it was only needed to support NGen

davidwrighton committed:
Merge branch 'main' of github.com:dotnet/runtime into faster_type_load

davidwrighton committed:
Fix up some more issues

davidwrighton committed:
Bring back late virtual propagation in the presence of covariant overrides only

davidwrighton committed:
Check correct flag on EEClass

davidwrighton committed:
Drive by fix for GetRestoredSlot. We don't need the handling of unrestored slots anymore

davidwrighton committed:
Fix composite build with new tables

davidwrighton committed:
Uniquify the mangled names

davidwrighton commented:
/azp run runtime-coreclr crossgen2

davidwrighton committed:
Add more __

davidwrighton committed:
Initial pass at type skip verifation checker

davidwrighton committed:
Fix logging and some correctness issues

davidwrighton committed:
Enable the more of type checking
- Notably, the recursive stuff now works
- Also fix a bug in constraint checking involving open types in the type system

AaronRobinsonMSFT commented on src/coreclr/inc/readytorun.h:
```diff
@@ -92,6 +92,9 @@ enum class ReadyToRunSectionType : uint32_t
     ManifestAssemblyMvids       = 118, // Added in V5.3
     CrossModuleInlineInfo       = 119, // Added in V6.2
     HotColdMap                  = 120, // Added in V8.0
+    MethodIsGenericMap          = 121, // Added in V9.0
+    EnclosingTypeMap               = 122, // Added in V9.0
+    TypeGenericInfoMap          = 123, // Added in V9.0
```
```suggestion
    MethodIsGenericMap          = 121, // Added in V9.0
    EnclosingTypeMap            = 122, // Added in V9.0
    TypeGenericInfoMap          = 123, // Added in V9.0
```

AaronRobinsonMSFT commented on src/coreclr/inc/readytorun.h:
```diff
@@ -121,6 +124,21 @@ enum class ReadyToRunImportSectionFlags : uint16_t
     PCode    = 0x0004, // Section contains pointers to code
 };
 
+enum class ReadyToRunTypeGenericInfo : uint8_t
```
Is there a reason these are the values they are?

davidwrighton replied:
Yes, by choosing these values we are able to store the data here in 4 bit chunks which pack nicely into 8 bit bytes.

AaronRobinsonMSFT commented on src/coreclr/vm/readytoruninfo.h:
```diff
@@ -48,6 +48,42 @@ class ReadyToRunCoreInfo
+{
+    uint32_t MethodCount = 0;
+public:
+    static ReadyToRun_MethodIsGenericMap EmptyInstance;
```
Can all these `EmptyInstance` values be marked `const` or is that going to mess with things?

AaronRobinsonMSFT commented on src/coreclr/vm/readytoruninfo.h:
```diff
@@ -48,6 +48,42 @@ class ReadyToRunCoreInfo
 typedef DPTR(class ReadyToRunCoreInfo) PTR_ReadyToRunCoreInfo;
 
+
+class ReadyToRun_EnclosingTypeMap
```
If I'm reading how this works correctly, these are partial types, right? As in they represent a collection and are cast to a pointer and then indexed into. I get this from looking at the implementation of `GetTypeGenericInfo()`.

My suggestion here would be not make the type that has the logic be the data too. Instead create a struct that is the data and then use it. For example, I think for `ReadyToRun_TypeGenericInfoMap` the data generally looks like the following:

```cpp
struct ReadyToRun_TypeGenericInfoMap_Data
{
    uint32_t TypeCount;
    uint8_t Data[ANYSIZE_ARRAY];
};
```

The logic for using the index from a RID makes sense, but then you check the lower bit after dividing it by 2. Is this an odd/even thing? I will comment there.

davidwrighton replied:
Seems like a good plan, to be honest, this code ended up kinda gross.

AaronRobinsonMSFT commented on src/coreclr/vm/readytoruninfo.cpp:
```diff
@@ -1771,5 +1789,161 @@ ModuleBase* CreateNativeManifestModule(LoaderAllocator* pLoaderAllocator, IMDInt
+        return 0;
+    }
+
+    uint8_t chunk = ((uint8_t*)&MethodCount)[((rid - 1) / 8) + 4];
+    chunk >>= 7 - ((rid - 1) % 8);
+    return !!(chunk & 1);
```
Where is this data layout documented?

AaronRobinsonMSFT commented on src/coreclr/vm/readytoruninfo.cpp:
```diff
@@ -1771,5 +1789,161 @@ ModuleBase* CreateNativeManifestModule(LoaderAllocator* pLoaderAllocator, IMDInt
+        *foundResult = false;
+        return (ReadyToRunTypeGenericInfo)0;
+    }
+    uint32_t index = rid - 1;
+    uint8_t *pTypeGenericInfo = ((uint8_t*)&TypeCount) + sizeof(uint32_t);
+    uint8_t entry = pTypeGenericInfo[index / 2];
+    if (!(index & 1))
+    {
+        entry >>= 4;
+    }
```
Where is this data layout documented?

davidwrighton replied:
Nowhere, yet. That was next on my list of things to do once I had my final perf numbers to put into the checkin comment, and proved that it worked fairly completely. I should have that data by end of night tonight, and then I'll work on fixing up the command line for controlling this stuff, and adding some docs.

davidwrighton committed:
Fix build breaks involving new feature of GenericParameterDesc

davidwrighton committed:
Merge branch 'main' of github.com:dotnet/runtime into faster_type_load

davidwrighton commented:
/azp run runtime-coreclr crossgen2

davidwrighton committed:
Add documentation for R2R format changes
Fix command line parameter to be more reasonable, and allow logging on command
Fix the rest of issues noted in crossgen2 testing

AaronRobinsonMSFT commented on docs/design/coreclr/botr/readytorun-format.md:
```diff
@@ -644,6 +648,33 @@ The entry of the hashtable is a counted sequence of compressed unsigned integers
 This section may be included in addition to a InliningInfo2 section.
 
+## ReadyToRunSectionType.MethodIsGenericMap (v9.0+)
+This optional section holds a bit vector to indicate if the MethodDef's contained within the assembly have generic parameters or not. This allowes determinining if a method is generic or not by querying a bit vector (which is fast, and efficient) as opposed to examining the GenericParameter table, or the signature of the Method.
```
```suggestion
This optional section holds a bit vector to indicate if the MethodDefs contained within the assembly have generic parameters or not. This allows determining if a method is generic or not by querying a bit vector (which is fast, and efficient) as opposed to examining the GenericParameter table, or the signature of the Method.
```

AaronRobinsonMSFT commented on docs/design/coreclr/botr/readytorun-format.md:
```diff
@@ -644,6 +648,33 @@ The entry of the hashtable is a counted sequence of compressed unsigned integers
+## ReadyToRunSectionType.MethodIsGenericMap (v9.0+)
+This optional section holds a bit vector to indicate if the MethodDef's contained within the assembly have generic parameters or not. This allowes determinining if a method is generic or not by querying a bit vector (which is fast, and efficient) as opposed to examining the GenericParameter table, or the signature of the Method.
+
+The section begins with a single 32bit integer indicating the number of bits in the bit vector. Following that integer is the actual bit vector of all of the data. The data is grouped into 8 bit bytes, where the least significant bit of the byte is the bit which represents the lowest MethodDef. 
```
```suggestion
The section begins with a single 32 bit integer indicating the number of bits in the bit vector. Following that integer is the actual bit vector of all of the data. The data is grouped into 8 bit bytes, where the least significant bit of the byte is the bit which represents the lowest MethodDef.
```

AaronRobinsonMSFT commented on docs/design/coreclr/botr/readytorun-format.md:
```diff
@@ -644,6 +648,33 @@ The entry of the hashtable is a counted sequence of compressed unsigned integers
+This optional section represents a condensed view of some generic details about types. This can make it more efficient to load types.
+
+The structure of this section is:
+A single 32bit integer representing the number of entries in the map followed by a series of 4 bit entries, one per type. These 4 bit entries are grouped into bytes, where each byte holds 2 entries, and the entry in the most signficant 4 bits of the byte is the entry representing a lower TypeDef RID.
```
```suggestion
A single 32 bit integer representing the number of entries in the map followed by a series of 4 bit entries, one per type. These 4 bit entries are grouped into bytes, where each byte holds 2 entries, and the entry in the most significant 4 bits of the byte is the entry representing a lower TypeDef RID.
```

AaronRobinsonMSFT commented on src/coreclr/tools/Common/TypeSystem/Common/GenericParameterDesc.cs:
```diff
@@ -87,6 +87,12 @@ public virtual string Name
+        /// </summary>
+        public abstract TypeSystemEntity AssociatedTypeOrMethod { get; }
+
+
```
```suggestion
```

AaronRobinsonMSFT commented on src/coreclr/tools/aot/ILCompiler.ReadyToRun/Compiler/DependencyAnalysis/ReadyToRun/EnclosingTypeMapNode.cs:
```diff
@@ -0,0 +1,70 @@
+            return ((typeDefinitionCount > 10) && (typeDefinitionCount <= 0xFFFE));
+        }
+
+        public override int ClassCode => 990540812;
```
What is this value?

davidwrighton replied:
Its a random number. It needs to be different from other random numbers used as ClassCode, but that's the only requirement.

AaronRobinsonMSFT commented on src/coreclr/tools/aot/ILCompiler.ReadyToRun/Compiler/DependencyAnalysis/ReadyToRun/EnclosingTypeMapNode.cs:
```diff
@@ -0,0 +1,70 @@
+            ObjectDataBuilder builder = new ObjectDataBuilder(factory, relocsOnly);
+            builder.AddSymbol(this);
+
+            // This map is only valid for assemblies with <= 0xFFFE types defined within
```
I would assert this invariant.

davidwrighton replied:
Done

AaronRobinsonMSFT commented on src/coreclr/tools/aot/ILCompiler.ReadyToRun/Compiler/DependencyAnalysis/ReadyToRun/MethodIsGenericMapNode.cs:
```diff
@@ -0,0 +1,68 @@
+            ObjectDataBuilder builder = new ObjectDataBuilder(factory, relocsOnly);
+            builder.AddSymbol(this);
+
+            builder.EmitInt(_metadata.MethodDefinitions.Count);
```
Assert 32 limit.

davidwrighton replied:
Structurally this can't be a problem. Count is limited by the ECMA format to 24bits.

AaronRobinsonMSFT commented on src/coreclr/tools/aot/ILCompiler.ReadyToRun/Compiler/DependencyAnalysis/ReadyToRun/TypeGenericInfoMapNode.cs:
```diff
@@ -0,0 +1,105 @@
+            ObjectDataBuilder builder = new ObjectDataBuilder(factory, relocsOnly);
+            builder.AddSymbol(this);
+
+            builder.EmitUInt((uint)_metadata.TypeDefinitions.Count);
```
Assert the `IsSupported()`?

davidwrighton commented on src/coreclr/vm/method.hpp:
```diff
@@ -144,27 +144,28 @@ enum MethodDescClassification
 
-    // unused                           = 0x0100,
-    // unused                           = 0x0200,
+    mdcValueTypeParametersWalked        = 0x0100, // Indicates that all typeref's in the signature of the method have been resolved
```
It is a significant performance improvement; but yeah, its a lot of bits. If you squint there are ways to get a few more bits out of the system if we need them without significant cost. There are other ways to fix the perf problem here, but they aren't as general, and involve more significant refactoring. At the moment this looks like the right choice, but if we find we really need 4 or 5 more bits for something else, I have some ideas on how to go about that.

davidwrighton committed:
Fix implementation of CompareMethodContraints. instead of using IsGeneric map, check to see if the method is generic in the first place. It turns out we have an efficient way to check in every place that matters

davidwrighton committed:
Fix nits noticed by Aaron

davidwrighton committed:
Add some const correctness to the world

MichalStrehovsky commented on src/coreclr/nativeaot/System.Private.TypeLoader/src/Internal/TypeSystem/RuntimeGenericParameterDesc.cs:
```diff
@@ -25,5 +26,12 @@ public RuntimeGenericParameterDesc(GenericParameterKind kind, int index, TypeSys
+
+        public override TypeSystemEntity AssociatedTypeOrMethod => _associatedTypeOrMethod;
+
+        internal void SetAssociatedTypeOrMethod(TypeSystemEntity associatedTypeOrMethod)
+        {
+            _associatedTypeOrMethod = associatedTypeOrMethod;
+        }
```
Can we do this in a constructor? This is already a `RuntimeXXXDesc` so I wouldn't mind if the constructor just understands a `ReadOnlySpan` of `Runtime.GenericVariance`.

davidwrighton replied:
Refactor requested and implemented.

MichalStrehovsky commented on src/coreclr/tools/Common/TypeSystem/Ecma/EcmaGenericParameter.cs:
```diff
@@ -78,6 +79,22 @@ public override GenericParameterKind Kind
+        {
+            get
+            {
+                TypeSystemEntity tse = Volatile.Read(ref _associatedTypeOrMethod);
+                if (tse == null)
+                {
+                    GenericParameter parameter = _module.MetadataReader.GetGenericParameter(_handle);
+
+                    tse = (TypeSystemEntity)_module.GetObject(parameter.Parent);
+                    Volatile.Write(ref _associatedTypeOrMethod, tse);
+                }
+                return tse;
```
I don't think we need the Volatile here.

If this is perf critical, we can use the same pattern as BaseType:

https://github.com/dotnet/runtime/blob/8ee61feba4f0d906d0ceeaea15b622f9b078c87f/src/coreclr/tools/Common/TypeSystem/Ecma/EcmaType.cs#L174-L182

If it's not perf critical maybe we don't even need to cache it.

MichalStrehovsky commented on src/coreclr/tools/ILVerification/InstantiatedGenericParameter.cs:
```diff
@@ -76,6 +76,8 @@ private InstantiatedGenericParameter(GenericParameterDesc genericParam)
 
         public override GenericConstraints Constraints => _genericParam.Constraints;
 
+        public override TypeSystemEntity AssociatedTypeOrMethod => _genericParam.AssociatedTypeOrMethod;
```
I don't know much about the ILVerify codebase, but doesn't this need `InstantiateSignature`?

davidwrighton replied:
Maybe? I don't know that codebase either, but it doesn't seem unreasonable.

davidwrighton commented on src/coreclr/tools/Common/TypeSystem/Common/SignatureVariable.cs:
```diff
@@ -66,6 +66,10 @@ protected override TypeFlags ComputeTypeFlags(TypeFlags mask)
                 flags |= TypeFlags.SignatureTypeVariable;
             }
 
+            flags |= TypeFlags.HasGenericVarianceComputed;
+            flags |= TypeFlags.HasFinalizerComputed;
+            flags |= TypeFlags.AttributeCacheComputed;
```
Unfortunately, this work operates almost entirely over uninstantiated types and methods, and has a few spots where it actually needs to use these things. However, it might be the case that I can only set a minimal number of flags. I enabled them all as it seemed like the right thing to do, but I believe the only 1 I really needed to work was the attribute cache.

MichalStrehovsky replied:
What's the stack that leads to this being needed? This really feels like something where the only right answer is to unask the question.

davidwrighton committed:
Fix issues noted by Michal, as well as remaining constrain checking issues

AaronRobinsonMSFT commented on src/coreclr/vm/readytoruninfo.h:
```diff
@@ -48,6 +48,42 @@ class ReadyToRunCoreInfo
+    ReadyToRunTypeGenericInfo GetTypeGenericInfo(mdTypeDef input, bool *foundResult) const;
+public:
+    const static ReadyToRun_TypeGenericInfoMap EmptyInstance;
+
```
```suggestion
    ReadyToRun_TypeGenericInfoMap() = delete;
```

AaronRobinsonMSFT commented on src/coreclr/vm/readytoruninfo.h:
```diff
@@ -48,6 +48,42 @@ class ReadyToRunCoreInfo
+class ReadyToRun_MethodIsGenericMap
+{
+    uint32_t MethodCount = 0;
+public:
```
```suggestion
public:
    ReadyToRun_MethodIsGenericMap() = delete;
```

AaronRobinsonMSFT commented on src/coreclr/vm/readytoruninfo.h:
```diff
@@ -48,6 +48,42 @@ class ReadyToRunCoreInfo
+    uint16_t TypeCount = 0;
+public:
+    const static ReadyToRun_EnclosingTypeMap EmptyInstance;
+
```
```suggestion
    ReadyToRun_EnclosingTypeMap() = delete;
```

davidwrighton committed:
Merge branch 'main' of github.com:dotnet/runtime into faster_type_load

davidwrighton committed:
Code review details

davidwrighton committed:
Merge branch 'main' of github.com:dotnet/runtime into faster_type_load

davidwrighton committed:
Merge branch 'main' into faster_type_load

AaronRobinsonMSFT reviewed (approved):
The VM side looks good to me.

davidwrighton commented:
@trylek  Could you take a look at the Crossgen2 side of these changes?

trylek commented on src/coreclr/tools/aot/ILCompiler.ReadyToRun/Compiler/DependencyAnalysis/ReadyToRun/EnclosingTypeMapNode.cs:
```diff
@@ -0,0 +1,71 @@
+            builder.AddSymbol(this);
+
+            // This map is only valid for assemblies with <= 0xFFFE types defined within
+            Debug.Assert(_metadata.TypeDefinitions.Count <= 0xFFFE);
```
Nit - this constant appears multiple times in the code, could we make it a symbolic constant? A sufficiently expressive name like <code>MaxTypeCountSupportingEnclosingTypeMap</code> could also make some of the code comments obsolete / self-explanatory.

davidwrighton replied:
Good idea.

trylek commented on src/coreclr/tools/aot/ILCompiler.ReadyToRun/Compiler/DependencyAnalysis/ReadyToRun/HeaderNode.cs:
```diff
@@ -94,7 +150,23 @@ public override ObjectData GetData(NodeFactory factory, bool relocsOnly = false)
+                {
+                    if (factory.OptimizationFlags.TypeValidation == TypeValidationRule.AutomaticWithLogging)
+                    {
+                        foreach (string reason in _shouldAddSkipTypeValidationFlag.Result.reasons)
```
I might be missing part of the intended logic here - according to the naming I would expect <code>_shouldAddSkipTypeValidationFlag</code> to represent some form of vetting whether we can drop the type validation and demonstrate the rationale if we do; why do we dump the reasons in the <code>else</code> block instead?

davidwrighton replied:
Added a comment. These are logging reasons why we aren't able to skip validation. Logging reasons we could skip would be really boring. (It would just be a list of types that are considered ok, which isn't all that useful for anybody)

trylek commented on src/coreclr/tools/aot/ILCompiler.ReadyToRun/Compiler/DependencyAnalysis/ReadyToRun/HeaderNode.cs:
```diff
@@ -33,6 +35,50 @@ public override ObjectNodeSection GetSection(NodeFactory factory)
         }
     }
 
+    public abstract class ModuleSpecificHeaderTableNode : HeaderTableNode
```
I really like your introduction of the <code>ModuleSpecificHeaderTableNode</code>, that makes the code much easier to understand.

davidwrighton replied:
Thank you.

trylek commented on src/coreclr/vm/readytoruninfo.cpp:
```diff
@@ -1771,5 +1789,161 @@ ModuleBase* CreateNativeManifestModule(LoaderAllocator* pLoaderAllocator, IMDInt
+    return count;
+}
+
+HRESULT ReadyToRun_TypeGenericInfoMap::GetGenericArgumentCountNoThrow(mdTypeDef input, uint32_t *pCount, IMDInternalImport* pImport) const
```
Nit - there seems to be tons of duplication between the pairs of methods with / without the <code>NoThrow</code> prefix. Could we somehow consolidate them by making one call the other and just slightly tweak the outputs?

trylek commented on src/coreclr/vm/readytoruninfo.cpp:
```diff
@@ -1771,5 +1789,161 @@ ModuleBase* CreateNativeManifestModule(LoaderAllocator* pLoaderAllocator, IMDInt
+        return 0;
+    }
+
+    uint8_t chunk = ((uint8_t*)&MethodCount)[((rid - 1) / 8) + 4];
```
Nit - if you want to be consistent with your implementation of the other tables (and possibly also make the code easier to understand), you might want to consider replacing 4 with <code>sizeof(uint32_t)</code> to make it obvious it represents the initial count entry in the table.

trylek reviewed (approved):
Looks great to me, thanks David for making all these optimizations!

davidwrighton committed:
Code review from trylek

davidwrighton committed:
Merge branch 'main' of https://github.com/dotnet/runtime into faster_type_load
